api/query_test.go

```diff
@@ -0,0 +1,330 @@
+		},
+		{
+			name:   "Missing parameters GET query",
+			method: http.MethodGet,
+			sliceQuery: "{\"vds\":\"" + well_known +
+				"\", \"direction\":\"i\", \"sas\": \"n/a\"}",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "Error:Field validation for 'Lineno'",
+		},
+		{
+			name:   "Missing parameters POST Query",
+			method: http.MethodPost,
+			sliceQuery: "{\"vds\":\"" + well_known +
+				"\", \"lineno\":1, \"sas\": \"n/a\"}",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "Error:Field validation for 'Direction'",
+		},
+		{
+			name:   "Query with unknown axis",
+			method: http.MethodPost,
+			slice: TestSliceQuery{
+				Vds:       well_known,
+				Direction: "unknown",
+				Lineno:    1,
+				Sas:       "n/a",
+			},
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "Invalid direction 'unknown', valid options are",
+		},
+		{
+			name:   "Query which passed all input checks but still should fail",
+			method: http.MethodPost,
+			slice: TestSliceQuery{
+				Vds:       well_known,
+				Direction: "i",
+				Lineno:    10,
+				Sas:       "n/a",
+			},
+			expectedStatus: http.StatusInternalServerError,
+			expectedError:  "Invalid lineno: 10, valid range: [0:2:1]",
+		},
+	}
+
+	for _, testcase := range testcases {
+		w := httptest.NewRecorder()
+		ctx, r := gin.CreateTestContext(w)
+		r.Use(gin.Logger())
+		r.Use(ErrorHandler)
+
+		endpoint := Endpoint{StorageURL: "", Protocol: "file://"}
+
+		r.POST("/slice", endpoint.SlicePost)
+		r.GET("/slice", endpoint.SliceGet)
+
+		prepareSliceRequest(ctx, t, testcase)
+		r.ServeHTTP(w, ctx.Request)
+
+		if w.Result().StatusCode != testcase.expectedStatus {
+			msg := "Got status %v; want %d %s in case '%s'"
+			t.Errorf(
+				msg,
+				w.Result().Status,
+				testcase.expectedStatus,
+				http.StatusText(testcase.expectedStatus),
+				testcase.name,
+			)
+		}
+		if testcase.expectedError == "" {
+			parts := readMultipartData(t, w)
+			if len(parts) != 2 {
+				msg := "Got %d parts in reply; want it to always contain 2 in case '%s'"
+				t.Errorf(
+					msg,
+					len(parts),
+					testcase.name,
+				)
+			}
+
+			type SliceAxis struct {
+				Annotation string  `json:"annotation" binding:"required"`
+				Max        float32 `json:"max"        binding:"required"`
+				Min        float32 `json:"min"        binding:"required"`
+				Samples    int     `json:"samples"    binding:"required"`
+				Unit       string  `json:"unit"       binding:"required"`
+			}
+
+			type SliceMetadata struct {
+				Axis   []SliceAxis `json:"axis"   binding:"required"`
+				Format int         `json:"format" binding:"required"`
+			}
+
+			inlineAxis := SliceAxis{
+				Annotation: "Inline", Max: 5.0, Min: 1.0, Samples: 3, Unit: "unitless",
+			}
+			crosslineAxis := SliceAxis{
+				Annotation: "Crossline", Max: 11.0, Min: 10.0, Samples: 2, Unit: "unitless",
+			}
+			sampleAxis := SliceAxis{
+				Annotation: "Sample", Max: 16.0, Min: 4.0, Samples: 4, Unit: "ms",
+			}
```

Comment:
Testing stuff like this should be tested in `vds_test.go` instead I think. Here we should only test the correspondence between metadata and data. I.e. the data length thing at the bottom here

Reply:
`vds_test.go` should properly test that, if there are some unexpected/different cases or whatever.
 
But you need to happy-test it once here too (to make sure that metadata is as expected, no json stuff was lost in query.go get/parse or in the middleware), as we consider these our most outer tests run on every change.

I like to indirectly "randomize" parameters a bit (do not use the same value everywhere), but I again forgot that with removal of form I removed half of the cases where this randomization was used.
Removed unused to make stuff below a bit simpler.

Reply:
I'm arguing that `expectedDataLength := expectedMetadata.Axis[0].Sample` is probably giving us enough assurance that the metadata is correct. After all, nothing really happens with the metadata in the http layer, other than writing a buffer. 

But if you insists all of this setup out-benefits the noise it creates, then I leave it alone 😄